src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,270 @@
+        return isEmpty() ? Optional.empty() : Optional.of(this);
+    }
+
+    /**
+     * Stream codec for an item resource. Accepts empty resources.
+     */
+    public static final StreamCodec<RegistryFriendlyByteBuf, ItemResource> STREAM_CODEC = StreamCodec.composite(
+            ByteBufCodecs.holderRegistry(Registries.ITEM), ItemResource::getHolder,
+            DataComponentPatch.STREAM_CODEC, ItemResource::getComponentsPatch,
+            ItemResource::of);
+
+    public static ItemStack itemStackOf(IResourceStack<ItemResource> resourceStack) {
+        return resourceStack.resource().toStack(resourceStack.amount());
+    }
+
+    public static final ItemResource EMPTY = new ItemResource(ItemStack.EMPTY);
+    public static final ResourceStack<ItemResource> EMPTY_STACK = ResourceStack.of(EMPTY, 0);
+    public static final MutableResourceStack<ItemResource> EMPTY_MUTABLE_STACK = MutableResourceStack.of(EMPTY, 0);
+
+    /**
+     * This is used only for registry, you should not use this method!
+     */
+    @ApiStatus.Internal
+    public static ItemResource invalidateDefault(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return new ItemResource(item.asItem().getDefaultInstance().copyWithCount(1));
+    }
+
+    public static ItemResource of(ItemStack itemStack) {
+        if (itemStack.isEmpty()) return EMPTY;
+
+        if (itemStack.isComponentsPatchEmpty())
+            return itemStack.getItem().defaultResource();
+
+        return new ItemResource(itemStack.copyWithCount(1));
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     */
+    public static ItemResource of(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return item.asItem().defaultResource();
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     */
+    public static ItemResource of(Holder<Item> item, DataComponentPatch patch) {
+        if (item.value() == Items.AIR) return EMPTY;
+        return item.value().defaultResource().withPatch(patch);
+    }
+
+    /**
+     * We wrap an item stack which must never be modified.
+     */
+    final ItemStack innerStack;
+
+    private ItemResource(ItemStack innerStack) {
+        this.innerStack = innerStack;
+    }
+
+    @Override
+    public Item getInstanceValue() {
+        return innerStack.getItem();
+    }
+
+    @Override
+    public Holder<Item> getHolder() {
+        return innerStack.getItemHolder();
+    }
+
+    //Defers to the stack in case there are injections done to it at some point
+    @Override
+    public boolean isEnabled(FeatureFlagSet enabledFeatures) {
+        return innerStack.isItemEnabled(enabledFeatures);
+    }
+
+    @Nullable
+    public <T, C extends @Nullable Object> T getCapability(ItemCapability<T, C> capability, C context) {
+        return capability.getCapability(toStack(), context);
+    }
```

Comment:
Seems very unnecessary, especially if the capability requires a mutable stack for whatever reason.